### ISSUE TITLE
Emoji: Skin tone modifiers used in isolation

### ISSUE THREAD

**Issue by janlelis** [jquast/wcwidth#133]
Hi Jeff,

in the wcwidth specification you list [skin tone modifiers](https://wcwidth.readthedocs.io/en/latest/specs.html) as being zero-width. However, this is not always true, since they should be displayed as Emoji (2 columns) when used in isolation / not part of a (known) Emoji sequence:

https://www.unicode.org/reports/tr51/#def_basic_emoji_set

Related: #134

**Comment by jquast** [comment]
Yes. There are combining characters that also have this exact same behavior.

This is difficult because it would mean having a definition of, "0 width when in sequence with [... large list ...], width 2 otherwise", and tracking every likely combination. 

I think downstream developers would all wish to have a faster wcwidth measurement at the cost of unable to measure these edge cases. Only with emoji is this list somewhat possible, (emoji-variation-seuqneces.txt), I trialed a small implementation of this, but it seemed like a very high cost of resources/slower measurement for the ~0.1% use case!

I have decided for emoji and combining characters that have 0 or 2 widths, depending on their cojoined character in sequence, that wcwidth expects only that they are used in their most common and combined form, and their uncommon standalone form is not supported.

This means that if a developer wishes to display a single fitzpatrick skin tone emoji or a hangul jungseong (korean combining character, example test https://github.com/jquast/wcwidth/blob/57cfbda2a87d9d6c5a2ba7b5018f3ce2c04b6383/tests/test_core.py#L225-L252), that they should not rely on python wcwidth library to measure the width.

**Comment by janlelis** [comment]
Thanks for the explanations and example! I think I agree that it is unfeasible to detect which combining combinations are allowed and which are not.

On the other hand I'd say that Emoji skin tone modifiers (and maybe also Hangul characters, but I have to understand them more) are different from most other combining marks in that they are supposed to have a **meaningful** standalone form, for example, in an Emoji picker UI widget (which I would not consider an edge case). Codepoints like [U+0308](https://codepoints.net/U+0308) don't have such a mode at all, they always "combine".

While there are many aspects of the Emoji standard that are pretty wild, it is fortunately clearly defined which Emoji characters can be combined with skin tone modifiers: Only those that have the Emoji modifier base property in the current Emoji data file:

- https://www.unicode.org/reports/tr51/#def_emoji_modifier_base
- https://www.unicode.org/Public/16.0.0/ucd/emoji/emoji-data.txt <-- codepoints with "Emoji_Modifier_Base"

Since this combination mechanism only involves the preceding codepoint, it should be possible to implement it without a large performance impact.
